File: go/anagram/anagram.go
// Package anagram is a rearrangement of letters to form a new word.
package anagram

import (
	"sort"
	"strings"
)

// Detect returns a string slice anagram of the passed string.
func Detect(subject string, candidates []string) []string {
	var anagrams []string = []string{}
	var tmpSlice []string = []string{}
	var subjectLower string
	var candidateLower string

	subjectLower = strings.ToLower(subject)
	tmpSlice = strings.Split(subjectLower, "")
	sort.Strings(tmpSlice)
	sortedWord := strings.Join(tmpSlice, "")

	for _, candidate := range candidates {
		candidateLower = strings.ToLower(candidate)
		tmpSlice = strings.Split(candidateLower, "")
		sort.Strings(tmpSlice)
		sortedCandidate := strings.Join(tmpSlice, "")

		if sortedWord == sortedCandidate && subjectLower != candidateLower {
			anagrams = append(anagrams, candidate)
		}
	}

	return anagrams
}

File: dart/anagram/lib/anagram.dart
// Find anagrams for a word from a list of candidates.
// https://exercism.org/tracks/dart/exercises/anagram

typedef word_t = String;
typedef letter_t = String;
typedef candidates_t = List<word_t>;
typedef matches_t = List<word_t>;
typedef letters_t = List<letter_t>;

class Anagram {
  matches_t findAnagrams(word_t search_word, candidates_t candidates) {
    matches_t matches = <String>[];
    letters_t search_letters = search_word.split('');
    search_letters.sort();
    word_t search_term = search_letters.join('');
    search_term.toLowerCase();

    var iter = candidates.iterator;

    while (iter.moveNext()) {
      letters_t letters = iter.toString().split('');
      letters.sort();
      word_t candidate = letters.join('');

      if (search_term == candidate) {
        matches.add(candidate);
      }
    }

    matches.sort();

    return matches;
  }
}
